WIP Feature/6 0/Introducing Assessment Question Service

mstuder opened the pull request:
In this pull request you can continuously see the current status of the development of the assessment question service. The pull request is continuously updated and serves to document progress.

There will be a beta test from october 15th to the end of october. Until end of october it will be decided whether the assessment question service will be delivered. for ILIAS 6.

We provide here a continuously automatically updated test platform. If you are interested, please create an account and send me an e-mail so that I can give you the ILIAS administrator role:
http://feature-6-0-asq-v2.ilias-core.review.studer-raimann.ch

**Feature Wiki Page**
https://docu.ilias.de/goto.php?target=wiki_1357_Introducing_Assessment_Question_Service

**Initial concept**
* **2018-11-20 initial interface and database migration concept** see https://github.com/ILIAS-eLearning/ILIAS/pull/1183
https://github.com/bheyser/ILIAS/blob/9d74d279d85892c69a48b12f80543908decd109f/Services/AssessmentQuestion/docs/readme.md

**Technical discussions of new approaches**
Following technical new approaches were designed by _**Björn Heyser, Adrian Lüthi**_ and **_Martin Studer_**. Consulted with the SIG E-Assessment and discussed extensively with the Technical Board.

* **2019-05-27 ilias with cqrs components:** see https://github.com/ILIAS-eLearning/ILIAS/pull/1859 
Readme: https://github.com/studer-raimann/ILIAS/blob/feature/trunk/ilias_with_cqrs_components/vendor/ilias/IliasComponentCourse/README.md
* **2019-06-15 Assessment Question Service with the DDD and CQRS / Eventsourcing Pattern**: see https://github.com/ILIAS-eLearning/ILIAS/pull/1931
Readme: https://github.com/ILIAS-eLearning/ILIAS/blob/bd89c8194b07cc257abee749bea10b55b74e6f72/Services/AssessmentQuestion/Readme.md
* **2019-07-25 Feature/6 0/asq api change proposal:** see https://github.com/ILIAS-eLearning/ILIAS/pull/2016
Readme: https://github.com/ILIAS-eLearning/ILIAS/blob/5784e0a1b78a56cb5751d93dc824e6ffa3b51a24/Services/AssessmentQuestion/docs/readme.md
* **2019-07-28 Conceptual Discussion Feature/6 0/asq concept:** see https://github.com/ILIAS-eLearning/ILIAS/pull/2021/ 
**Final Concept:** https://github.com/ILIAS-eLearning/ILIAS/blob/7b261f43c6fe59938d72b84b270196089afbaa91/Services/AssessmentQuestion/docs/development/cqrs-and-event-sourcing-paradigm.md

kergomard commented:
Thank you @mstuder for this next step in the discussion of the new Assessment Question Service. As we looked at this PR today during our TB-Meeting we didn't exactly know what to do with it. We would kindly ask you to please expand the description to:
- Clearly state the purpose of it (why is it here)
- Where does it fit in the history of developing the Assessment Question Service
- What we can find in it
- Maybe some words on the current state
so that somebody finding it knows how to put it into context. If I understood you right last wee, you would like to "grow/improve" this PR regularly until the whole feature is developed, right? It might be good to have this stated clearly, too.
Thank you very much again!

Hadrian-II committed:
add form for matching question

Hadrian-II committed:
Merge branch 'feature/6-0/asq_v2' of https://git.studer-raimann.ch/ILIAS/Core into feature/6-0/asq_v2

Hadrian-II committed:
move add button into row

Hadrian-II committed:
add isComplete flag to question & questiondto

mstuder committed:
Merge branch 'feature/6-0/asq_v2' into 'feature/6-0/asq_v2_ICAQ-127'

Feature/6 0/asq v2

See merge request ILIAS/Core!312

Hadrian-II committed:
fix tinymce troubles

Hadrian-II committed:
remove debug prints

mstuder committed:
Merge branch 'feature/6-0/asq_v2' into 'feature/6-0/asq_v2_ICAQ-127'

Feature/6 0/asq v2

See merge request ILIAS/Core!314

mstuder committed:
Test run for fixed question set

mstuder committed:
vorbereitungen for projection of question results

mstuder committed:
fix create revision by save and return

mstuder committed:
fix formating scoreAndProjectTestAttempt

AluethII committed:
Merge branch 'feature/6-0/asq_v2_ICAQ-127' into 'feature/6-0/asq_v2'

Feature/6 0/asq v2 icaq 127

See merge request ILIAS/Core!315

mstuder committed:
removed ugly call back

AluethII committed:
Merge branch 'feature/6-0/asq_v2_revisions_by_set_online' into 'feature/6-0/asq_v2'

removed ugly call back

See merge request ILIAS/Core!316

Hadrian-II committed:
implement rendering of matching question

Hadrian-II committed:
implement match question gui interactions

Hadrian-II committed:
finished matching question implementation

mstuder committed:
persist user answer score in test

mstuder committed:
save Questions

Hadrian-II committed:
Merge remote-tracking branch 'origin/feature/6-0/asq_v2_persist_user_answers_in_test' into feature/6-0/asq_v2

Hadrian-II committed:
implement js for matchingquestion editor

Hadrian-II committed:
dont delete old projections on project and only project if changed

Hadrian-II committed:
implement iscomplete

Hadrian-II committed:
code cleanup

mstuder committed:
removed answer score interface

mstuder committed:
result processing

mstuder committed:
finished scoring view - first step

AluethII committed:
Merge branch 'feature/6-0/asq_v2_test_scoring' into 'feature/6-0/asq_v2'

Feature/6 0/asq v2 test scoring

See merge request ILIAS/Core!319

Hadrian-II committed:
Merge branch 'feature/6-0/asq_v2' of https://git.studer-raimann.ch/ILIAS/Core into feature/6-0/asq_v2

Hadrian-II committed:
fix iscomplete for editors

Hadrian-II committed:
update wip status

mstuder committed:
import questions and create them

mstuder committed:
import

mstuder committed:
Merge branch 'feature/6-0/asq_v2' into 'feature/6-0/asq_import'

# Conflicts:
#   Services/AssessmentQuestion/PublicApi/Authoring/AuthoringQuestion.php

mstuder committed:
Merge branch 'feature/6-0/asq_import' into 'feature/6-0/asq_v2'

Feature/6 0/asq import

See merge request ILIAS/Core!320

mstuder committed:
fixed bugs because of mergin

Hadrian-II committed:
skeleton implementation of essayquestion

Hadrian-II committed:
implemented form for essay question

Hadrian-II committed:
Merge branch 'feature/6-0/asq_v2' of https://git.studer-raimann.ch/ILIAS/Core into feature/6-0/asq_v2

Hadrian-II committed:
basic essay editor

Hadrian-II committed:
implement essay question scoring

Hadrian-II committed:
implement essayquestion

Hadrian-II committed:
show imagemap region selection popup as ilias popup thing

Hadrian-II committed:
finished matching question implementation

Hadrian-II committed:
update matching strings

Hadrian-II committed:
updated possible matchings in matching quesiton editor

Hadrian-II committed:
js cleanup

Hadrian-II committed:
fix typo in js

Hadrian-II committed:
show selected regions in imagequestion preview

Hadrian-II committed:
fix error in matching question

Hadrian-II committed:
image map create squares and circles on holding shift

Hadrian-II committed:
show existing shapes in preview

Hadrian-II committed:
use ctrl instead of shift for squares

Hadrian-II committed:
fix imagemap preview empty item

Hadrian-II committed:
implement limit answers for imagemap question

Hadrian-II committed:
save data on incomplete form

Hadrian-II committed:
switch singleline/multiline per JS

Hadrian-II committed:
remove generic form and rename question forms

Hadrian-II committed:
fix imagemap errors after creation of new line

Hadrian-II committed:
update method signatures in scoring

Hadrian-II committed:
fix errortextscoring

Hadrian-II committed:
fix handling of empty answers in errortext question

Hadrian-II committed:
dont ignore changes when only max_answers is changed in imagemap

Hadrian-II committed:
make image question reactive

nhaagen committed:
fix #26554: show installation-title in page header

fwolf-ilias committed:
Release 6.0 beta1

chfsx committed:
Merge branch 'trunk' into release_6

Hadrian-II committed:
add text editor for ordering question

klees committed:
Study Programme: users might be deleted already (#26735)

iszmais committed:
fixed regex replacement for old mysqls

iszmais committed:
abstracted bu saving

mjansenDatabay committed:
Merge pull request #2388 from iszmais/hotfix/certificate-migration-regex-replace6.0-stable

fixed regex replacement for old mysqls

klees committed:
Study Programme: apply coding style

klees committed:
Learning Sequence: apply coding style

klees committed:
Setup: apply coding style

klees committed:
Refinery: apply coding style

klees committed:
Data: apply coding style

alex40724 committed:
help slate should ignore context for state

alex40724 committed:
fix empty sec panel lists

Commit:
Fixing Disk-Quota

This ports the bug-fixes for mantis #26346 to ILIAS 6 and fixes a few
follow-up issues with the Disk Quota settings.

alex40724 committed:
fixed no items messages on dashboard

alex40724 committed:
fixed bug 0026769: Test6: Not possible to create a new wiki

alex40724 committed:
fixed bug #26770: Test6: Not possible to create a new mediacast

alex40724 committed:
fixed bug #26780: Test6: Opening a specific category throws error

catkrahl committed:
fix 0026594: mainbar entry icon lso

alex40724 committed:
fixed bug #26778: Test6: Survey administration crashes

alex40724 committed:
fixed bug #26777: Test6: File Admin - Disk quota → failed to open stream: No such file or directory

alex40724 committed:
fixed bug #26761: Test6: Opening Admin > Communication > Comments causes Whoops

alex40724 committed:
clarify badge assignment behaviour

catkrahl committed:
fix 0026595: worklflow listing in slate

gvollbach committed:
Code style/cleanup

alex40724 committed:
fixed #26558: Incomplete startingpoint list

alex40724 committed:
fixed bug #26794: Position 1 in Metabar for Help Glyph

alex40724 committed:
fixed bug #26787: Duplication of the first entry in the repository tree and the drilldown tree view

alex40724 committed:
prevent double listing of some items

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

klees committed:
Study Programme: fixed #26805

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
Merge branch 'release_6' of https://github.com/ILIAS-eLearning/ILIAS into release_6

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

mjansenDatabay committed:
Chatroom/Repository: Add fix for Mantis issue 26801

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

alex40724 committed:
coding style fixes

Commit:
Merge remote-tracking branch 'origin/release_6' into release_6

Hadrian-II committed:
show default title to make empty questions clickable

Hadrian-II committed:
improve tinyMCE handling

Hadrian-II committed:
translations for imageformpopup

mjansenDatabay committed:
Mail: Possible fix for Mantis issue 26818

mjansenDatabay committed:
Merge remote-tracking branch 'origin/release_6' into release_6

Hadrian-II committed:
fix descriptions for matching question

alex40724 committed:
fixed bug #26237: Load tags async after save has missing lang module

alex40724 committed:
fixed bug #26482: \ilObjSurvey::checkConstraint: count(): Parameter must be an array or an object that implements Countable

alex40724 committed:
fixed bug #26579: Test60: Competences cannot be opened (Whoops)

klees committed:
PSR-2 (+X) Big Bang

As announced on 2019-05-19 via developers@ and PR, this commit introduces the
[new ILIAS coding style](https://github.com/ILIAS-TechnicalBoard/ILIAS/blob/trunk_coding_style/docs/documentation/coding-style.md)
to release_6.

klees committed:
Setup: fixed #26784

Hadrian-II committed:
some polish for file question

alex40724 committed:
fixed bug #26407: Blog aus Übung anlegen liefert Whoops

alex40724 committed:
fixed bug #26413: Tagging in Kategories

Hadrian-II committed:
polish fileupload question

alex40724 committed:
show dates/context as description

alex40724 committed:
SystemFolder: Fix Mantis 26857

alex40724 committed:
fixed bug #26854: Erstellung von Zertifikaten im Arbeitsraum wirft Fehler

Hadrian-II committed:
simplify preview gui

Hadrian-II committed:
improve question preview handling

klees committed:
Study Programme: fix #26768

nhaagen committed:
fix 26559 - set titles for login-screen

mjansenDatabay committed:
Certificate: Fixed Mantis issue 26880

alex40724 committed:
Merge pull request #2396 from nhaagen/release_6

fix 26559,26554: Page-Titles

Amstutz committed:
UI Components: Fix #26884

Amstutz committed:
UI Components, Tests: Fix of #26885

Amstutz committed:
UI Components, Panel Action: Fix of #26842

Amstutz committed:
UI Components, Item: Fix of #26537

Amstutz committed:
UI Components, View Control: Improved the examples

alex40724 committed:
fixed bug #26550: Peer feedback PHP 7.2 Compatibility Issue

alex40724 committed:
fixed bug #26616: Last Visited with Milos-Icon

alex40724 committed:
fixed bug #26620: Slate Tags in Personal Workspace persists after globally deactivating the Tags-Service

Amstutz committed:
UI Components, Panel: Fixing Unit Tests

mjansenDatabay committed:
Certificate: Fix filename of certificate previews according to Mantis issue #26883

alex40724 committed:
fixed bug #26111: Cron job Reminders for Assignments sends a mail too early

mjansenDatabay committed:
Mail: Use type casts to be probably more robust against issues like 24961

mjansenDatabay committed:
Merge remote-tracking branch 'origin/release_6' into release_6

nhaagen committed:
fix 26757 - state preserved for tools button. (#2399)

This aligns the behavior of 'close slates' and 'toggle tools'

mjansenDatabay committed:
Mail: Fixed return type of replaced mail body

mjansenDatabay committed:
Merge remote-tracking branch 'origin/release_6' into release_6

mjansenDatabay committed:
Mail/NotificationCenter: Open inbox in same viewport

klees committed:
add all TestSuites again

klees committed:
reset WFE samples for code-gen to pre-PSR2

nhaagen committed:
fix 26748 - activate first tool, if tools are actually engaged, but the former tool went away

Commit:
Fix Problem with Listing in Section "Colour and Contrast

Amstutz committed:
Merge pull request #2403 from nhaagen/6.0r/ui/mb-tools-26757

 fix 26748 - activate first tool, if tools are actually engaged, but …

mjansenDatabay committed:
Mail: Change type of message from INFO to SUCCESS according to Mantis issue 26903

mjansenDatabay committed:
Mail: Open print action in new viewport according to Mantis issue 26902

catkrahl committed:
fix 26859: slate tree depth level bg-color

catkrahl committed:
fix 26851/26848: reduced space between tree entries

mjansenDatabay committed:
Mail/TextTemplates: Fixed Mantis issue 26909

catkrahl committed:
fix 26858: bulky engaged bg-color derived from brand primary

smeyer-ilias committed:
0026798: Failed test: Verfügbarkeit des Kurses einstellen

catkrahl committed:
fix 26817: footer height and centered text

catkrahl committed:
adjusted grey jstree expander

catkrahl committed:
fix 26859: slate tree depth level #2

smeyer-ilias committed:
0026687: „*“ als Mehrzeichen-Platzhaltersuche zeigt nicht alle Ergebnisse an und verändert den Suchradius

alex40724 committed:
fixed bug #17959: Deactivated chapters are visible on tree

alex40724 committed:
removed lagacy dashboard less file

alex40724 committed:
removed lagacy dashboard less file

nhaagen committed:
fix #26312 - metabar closes on outside click

catkrahl committed:
fix 26609: secondary panels line breaking and bg

catkrahl committed:
compiled delos.css

mjansenDatabay committed:
NotificationCenter: Add API method to close the NotificationCenter completely

mjansenDatabay committed:
NotificationCenter: Streamline descriptions of top items

Amstutz committed:
Merge pull request #2406 from nhaagen/6.0r/ui/metabar-26312

fix #26312 - metabar closes on outside click

alex40724 committed:
fixed bug #20571: missing lang var il_crs_member_45748

mjansenDatabay committed:
NotificationCenter: Add API method to close the NotificationCenter completely

mjansenDatabay committed:
NotificationCenter/OnScreenChat: Automatically close NotificationCenter if last conversation item is removed (according to Mantis issue 26910)

nhaagen committed:
UI: flickering of buttons, page jumps on slate engaged after reload (#26459, #26462)

* fix 26459, 26462 - flickering of buttons, page jumps on slate engaged after reload
*  fix 26459, 26462 - fix tests

mjansenDatabay committed:
OnScreenChat: Move creation of message UUID to module

mjansenDatabay committed:
OnScreenChat: Handle issue with missing history on multi-tab scenario when conversation is hidden in NotificationCenter

nhaagen committed:
fix #26312 - metabar closes on outside click - follow up

mjansenDatabay committed:
OnScreenChat: Handle issue with missing history on multi-tab scenario when conversation is hidden in NotificationCenter

Amstutz committed:
Merge pull request #2411 from nhaagen/6.0r/ui/metabar-26312

fix #26312 - metabar closes on outside click - follow up

Amstutz committed:
UI Components Meta Bar: #26953

Commit:
cloud fix: deep links

Hadrian-II committed:
reimplement formula question

smeyer-ilias committed:
0026971: Error Message when trying to access course

smeyer-ilias committed:
0026968: Failed test: Verwalten-Menü aufrufen

smeyer-ilias committed:
Fixed error_handling

Hadrian-II committed:
add strings for formula question

Commit:
fix typo in saml login form info

mjansenDatabay committed:
Certificate: Add template migration for new object types

Amstutz committed:
UI Components, Glyph: Fix of #26918

Amstutz committed:
UI Components, Glyph: Fix of #26916

Amstutz committed:
UI Components, Glyph: Fix of #26919

mjansenDatabay committed:
Certificate: Fix wrong fallthrough/no break

mjansenDatabay committed:
Setup/Forum: Fix creation of index (Mantis issue 26987)

Amstutz committed:
Globa Screen/Main Menu: Show Repository Link and Seperator (#2415)

* Globa Screen/Main Menu: Show Repository Link and Seperator

* Globa Screen/Main Menu: Show Repository Link and Seperator, fix docs

* Globa Screen/Main Menu: Show Repository Link and Seperator, removed seperator

Hadrian-II committed:
implement compound handling for formula scoring

chfsx committed:
moved class

tfamula committed:
Accessibility Control Concept, Fix 26725

smeyer-ilias committed:
Fixed 0026993: Uncaught PDOException: SQLSTATE[42000] in DB Update step 5521

smeyer-ilias committed:
Merge branch 'release_6' of https://github.com/ILIAS-eLearning/ILIAS into release_6

Hadrian-II committed:
polish for formula question

alex40724 committed:
fixed bug #21598: Fehlende Sprachvariable: -exc_members_already_assigned_team-

alex40724 committed:
fixed bug 24647: LSO icon is big in Adminstration / Repository / Respository object type

alex40724 committed:
fixed bug #26205: Card Title should not be displayed as a button, but as a link.

Hadrian-II committed:
Merge remote-tracking branch 'remotes/origin/release_6' into feature/6-0/asq_v2

Hadrian-II committed:
Merge branch 'release_6' into feature/6-0/asq_v2

Hadrian-II committed:
use original ilobjectgui file as merge failed

Hadrian-II committed:
make composer run again

Hadrian-II committed:
fix failed merge of ilobjtestgui

Hadrian-II committed:
make it run again

Hadrian-II committed:
rename pageobject used in asq from page to asqpageobject

Hadrian-II committed:
saner implementation of ilasqobjectgui

Hadrian-II committed:
fix page object creation

Hadrian-II committed:
additional cleanup

Hadrian-II committed:
initial cloze question data

Hadrian-II committed:
add gap configuration to gui

Hadrian-II committed:
implement cloze display and scoring

mstuder commented:
The Assessment Question Service is in a first review process https://github.com/studer-raimann/AssessmentQuestion

I close this pull request beacause this one here is obssolete. You find the current version of the assessment question service in this repository: https://github.com/studer-raimann/AssessmentQuestion
